ui-core/src/components/inputs/datePicker/utils.ts

```diff
@@ -0,0 +1,262 @@
+ * Spec 3: If the user clicks on a date that is before the currently selected start date, the new date becomes the start date and the previous start date becomes the end date.
+ * Spec 4: If the user clicks on a date that is after the currently selected start date, that date becomes the end date of the slot.
+ * Spec 5: If a slot is already defined (i.e., both start and end dates are defined) and the user clicks on a new date, the existing slot is cleared and the new date becomes the start date of the new slot.
+ */
+export function computeNewSelectedSlot(
+  clickedDate: Date,
+  prevSelectedSlot: CalendarSlot | undefined
+): CalendarSlot | undefined {
+  if (!prevSelectedSlot || prevSelectedSlot?.start === null) {
+    // Spec 1
+
+    return { start: clickedDate, end: null };
+  } else if (!prevSelectedSlot.end) {
+    if (normalizeDate(clickedDate).getTime() === normalizeDate(prevSelectedSlot.start).getTime()) {
+      // Spec 2
+      return undefined;
+    } else if (normalizeDate(clickedDate) < normalizeDate(prevSelectedSlot.start)) {
+      // Spec 3
+      return { start: clickedDate, end: prevSelectedSlot.start };
+    } else {
+      // Spec 4
+      return { start: prevSelectedSlot.start, end: clickedDate };
+    }
+  } else {
+    // Spec 5
+    return { start: clickedDate, end: null };
+  }
+}
+
+/**
+ * Format the selected slot to a string that can be displayed in the input field.
+ * @param slot The new selected slot
+ * @returns The formatted string
+ * Spec 0: If the slot is undefined, return an empty string.
+ * Spec 1: If the start and end dates are the same, return the formatted date of the start date.
+ * Spec 2: If the start and end dates are in the same year, return the formatted start date without the year and end date.
+ * Spec 3: If the start and end dates are in different years, return the formatted start date and end date with the year.
+ */
+export function formatSlotToInputValue(slot: CalendarSlot | undefined) {
+  let formattedDate = '';
+  if (slot) {
+    const { start, end } = slot;
+
+    if (isSameDay(start, end)) {
+      formattedDate = formatDateString(start);
+    } else {
+      let formattedStartDate = formatDateString(start);
+
+      if (start && end && start.getFullYear() === end.getFullYear()) {
+        formattedStartDate = formattedStartDate.slice(0, 5);
+      }
+
+      formattedDate = `${formattedStartDate} - ${formatDateString(end)}`;
+    }
+  }
+  return formattedDate;
+}
+
+/**
+ * Format the new value to a string that can be displayed in the input field.
+ * @param newValue The new value
+ * @param isRangeMode Whether the calendar is in range mode
+ * @returns The formatted string
+ */
+export function formatInputValue(
+  newValue: DatePickerProps['value'],
+  isRangeMode: DatePickerProps['isRangeMode']
+) {
+  if (isRangeMode) {
+    return formatSlotToInputValue(newValue as RangeDatePickerProps['value']);
+  }
+  return formatDateString(newValue as SingleDatePickerProps['value']);
+}
```

Comment:
It's a shame that we lose the typing here, but you can do it like this to avoid losing the type and casting:
```suggestion
export function formatInputValue({ value: newValue, isRangeMode }: DatePickerProps) {
  if (isRangeMode) {
    return formatSlotToInputValue(newValue);
  }
  return formatDateString(newValue);
}
```